dataplane/forwarding/fwdport/ports/kernel.go

```diff
@@ -0,0 +1,174 @@
+	if err := p.InitCounters("", "", list...); err != nil {
+		return nil, err
+	}
+	p.read()
+	return p, nil
+}
+
+type packetHandle interface {
+	gopacket.PacketDataSource
+	WritePacketData([]byte) error
+	Close()
+}
+
+type kernelPort struct {
+	fwdobject.Base
+	devName string
+	input   fwdaction.Actions
+	output  fwdaction.Actions
+	ctx     *fwdcontext.Context // Forwarding context containing the port
+	handle  packetHandle
+}
+
+func (p *kernelPort) String() string {
+	desc := fmt.Sprintf("Type=%v;DeviceName=%v;<Input=%v>;<Output=%v>", fwdpb.PortType_PORT_TYPE_KERNEL, p.devName, p.input, p.output)
+	if state, err := p.State(nil); err == nil {
+		desc += fmt.Sprintf("<State=%v>;", state)
+	}
+	return desc
+}
+
+func (p *kernelPort) Cleanup() {
+	p.input.Cleanup()
+	p.output.Cleanup()
+	p.handle.Close()
+	p.input = nil
+	p.output = nil
+}
+
+// Update updates a port.
+func (p *kernelPort) Update(upd *fwdpb.PortUpdateDesc) error {
+	var err error
+	defer func() {
+		if err != nil {
+			p.Cleanup()
+		}
+	}()
+	kernelUpd, ok := upd.Port.(*fwdpb.PortUpdateDesc_Kernel)
+	if !ok {
+		return fmt.Errorf("invalid type for port update")
+	}
+
+	// Acquire new actions before releasing the old ones.
+	if p.input, err = fwdaction.NewActions(kernelUpd.Kernel.GetInputs(), p.ctx); err != nil {
+		return fmt.Errorf("ports: input actions for port %v failed, err %v", p, err)
+	}
+	if p.output, err = fwdaction.NewActions(kernelUpd.Kernel.GetOutputs(), p.ctx); err != nil {
+		return fmt.Errorf("ports: output actions for port %v failed, err %v", p, err)
+	}
+	return nil
+}
+
+func (p *kernelPort) read() {
+	src := gopacket.NewPacketSource(p.handle, layers.LinkTypeEthernet)
+	go func() {
+		for {
+			select {
+			case pkt, ok := <-src.Packets():
+				if !ok {
+					log.Warning("src chan closed")
+					return
+				}
+				fwdPkt, err := fwdpacket.New(fwdpb.PacketHeaderId_PACKET_HEADER_ID_ETHERNET, pkt.Data())
+				if err != nil {
+					log.Warningf("failed to create new packet: %v", err)
+					fwdport.Increment(p, len(pkt.Data()), fwdpb.CounterId_COUNTER_ID_RX_BAD_PACKETS, fwdpb.CounterId_COUNTER_ID_RX_BAD_OCTETS)
+					continue
+				}
+				fwdport.Process(p, fwdPkt, fwdpb.PortAction_PORT_ACTION_INPUT, p.ctx, "Kernel")
+			}
+		}
+	}()
+}
+
+// Write writes a packet out. If successful, the port returns
+// fwdaction.CONSUME.
+func (p *kernelPort) Write(packet fwdpacket.Packet) (fwdaction.State, error) {
+	if err := p.handle.WritePacketData(packet.Frame()); err != nil {
+		return fwdaction.DROP, fmt.Errorf("failed to write eth packet: %v", err)
+	}
+	return fwdaction.CONSUME, nil
+}
+
+// Actions returns the port actions of the specified type
+func (p *kernelPort) Actions(dir fwdpb.PortAction) fwdaction.Actions {
+	switch dir {
+	case fwdpb.PortAction_PORT_ACTION_INPUT:
+		return p.input
+	case fwdpb.PortAction_PORT_ACTION_OUTPUT:
+		return p.output
+	}
+	return nil
+}
+
+// State manages the state of the port.
```

Comment:
returns or manages? I guess it's not using `op` right now, so is there a TODO?

Reply:
added todo